samples/s3-ls/s3-ls.c

```diff
@@ -0,0 +1,274 @@
+        int c = aws_cli_getopt_long(argc, argv, "r:v:h", s_long_options, &option_index);
+        if (c == -1) {
+            break;
+        }
+
+        switch (c) {
+            case 0:
+                /* getopt_long() returns 0 if an option.flag is non-null */
+                break;
+            case 'r':
+                ctx->region = aws_cli_optarg;
+                break;
+            case 'v':
+                if (!strcmp(aws_cli_optarg, "TRACE")) {
+                    ctx->log_level = AWS_LL_TRACE;
+                } else if (!strcmp(aws_cli_optarg, "INFO")) {
+                    ctx->log_level = AWS_LL_INFO;
+                } else if (!strcmp(aws_cli_optarg, "DEBUG")) {
+                    ctx->log_level = AWS_LL_DEBUG;
+                } else if (!strcmp(aws_cli_optarg, "ERROR")) {
+                    ctx->log_level = AWS_LL_ERROR;
+                } else {
+                    fprintf(stderr, "unsupported log level %s.\n", aws_cli_optarg);
+                    s_usage(1);
+                }
+                break;
+            case 'h':
+                s_usage(0);
+                break;
+            default:
+                fprintf(stderr, "Unknown option\n");
+                s_usage(1);
+        }
+    }
+
+    if (aws_cli_optind < argc) {
+        struct aws_byte_cursor uri_cursor = aws_byte_cursor_from_c_str(argv[aws_cli_optind++]);
+
+        if (aws_uri_init_parse(&ctx->uri, ctx->allocator, &uri_cursor)) {
+            fprintf(
+                stderr,
+                "Failed to parse uri %s with error %s\n",
+                (char *)uri_cursor.ptr,
+                aws_error_debug_str(aws_last_error()));
+            s_usage(1);
+        };
+    } else {
+        fprintf(stderr, "A URI for the request must be supplied.\n");
+        s_usage(1);
+    }
+}
+
+/**
+ * Predicate used to decide if the application is ready to exit.
+ * The corresponding condition variable is set when the last
+ * page of ListObjects is received.
+ */
+static bool s_app_completion_predicate(void *arg) {
+    struct app_ctx *app_ctx = arg;
+    return app_ctx->execution_completed;
+}
+
+/**
+ * Called once for each object returned in the ListObjectsV2 responses.
+ */
+bool s_on_object(const struct aws_s3_object_file_system_info *info, void *user_data) {
+    (void)user_data;
+
+    printf("%-18" PRIu64 " %.*s\n", info->size, (int)info->key.len, info->key.ptr);
+    return true;
+}
+
+/**
+ * Called once for each ListObjectsV2 response received.
+ * If the response contains a continuation token indicating there are more results to be fetched,
+ * requests the next page using aws_s3_paginator_continue.
+ */
+void s_on_list_finished(struct aws_s3_paginator *paginator, int error_code, void *user_data) {
+    struct app_ctx *app_ctx = user_data;
+
+    if (error_code == 0) {
+        bool has_more_results = aws_s3_paginator_has_more_results(paginator);
+
+        printf("______ has_more_results=%d\n", has_more_results);
```

Comment:
nit: good command line apps don't print to stdout information other than what another program could consume via. a pipe without a verbose option being specified.

Reply:
Removed, thanks!